content/docs/userspace/gall-guide/11-fail.md

```diff
@@ -0,0 +1,141 @@
++++
+title = "11. Failure"
+weight = 55
+template = "doc.html"
++++
+
+In this lesson we'll cover the last agent arm we haven't touched on yet:
+`on-fail`. We'll also touch on one last concept, which is the _helper core_.
+
+# Failures
+
+When crashes or errors occur in certain cases, Gall passes them to an agent's
+`on-fail` arm for handling. This arm is very seldom used, almost all agents
+leave it for `default-agent` to handle, which just prints the error message to
+the terminal. While you're unlikely to use this arm, we'll briefly go over its
+behavior for completeness.
+
+`on-fail` takes a `term` error message and a `tang`, typically containing a
+stack trace, and often with additional messages about the error. If it weren't
+delegated to `on-fail:def`, it would begin with:
+
+```hoon
+++  on-fail
+  |=  [=term =tang]
+  ^-  (quip card _this)
+  ....
+```
+
+Gall calls `on-fail` in four cases:
+
+- When there's a crash in the `on-arvo` arm.
+- When there's a crash in the `on-agent` arm.
+- When there's a crash in the `on-leave` arm.
+- When an agent produces a `%watch` card but the `wire`, ship, agent and `path`
+  specified are the same as an existing subscription.
+
+For an `on-arvo` failure, the `term` will always be `%arvo-response`, and the
+`tang` will contain a stack trace.
+
+For `on-agent`, the `term` will be the head of the `sign` (`%poke-ack`, `%fact`,
+etc). The `tang` will contain a stack trace and a message of "closing
+subscription".
```

Comment:
I wouldn't expect the stack trace to contain that in the case for `%poke-ack`, right?

Reply:
you wouldn't expect it but that's what the Gall source code said when I checked it. I just tested to confirm and it does in fact print "closing subscription" on any failed sign including poke-ack

Reply:
Hah, that's probably a gall bug then. Good catch!